src/cocotb/regression.py

```diff
@@ -176,109 +168,78 @@ def __init__(self, dut: SimHandle, tests: Iterable[Test]):
 
         self.xunit.add_property(name="random_seed", value=str(cocotb.RANDOM_SEED))
 
-        # Test Discovery
-        ####################
-        self._queue = []
-        for test in tests:
-            self.log.info(f"Found test {test.__module__}.{test.__qualname__}")
-            self._queue.append(test)
-        self.ntests = len(self._queue)
-
-        if not self._queue:
-            self.log.warning("No tests were discovered")
-
-        self._queue.sort(key=lambda test: (test.stage, test._id))
-
-    @classmethod
-    def from_discovery(cls, dut: SimHandle):
-        """
-        Obtains the test list by discovery.
+    def discover_tests(
+        self, modules: Iterable[str], filters: Optional[Sequence[str]] = None
+    ) -> None:
+        """Discovers tests in files automatically.
 
-        See :envvar:`MODULE` and :envvar:`TESTCASE` for details on how tests are discovered.
+        Should be called before :meth:`start_regression` is called.
+        If no filters are given, all tests are included.
+        If one or more filters are given, only those tests which match a filter are included; the rest are excluded.
 
         Args:
-            dut (SimHandle): The root handle to pass into test functions.
-        """
-        tests = cls._discover_tests()
-        return cls(dut, tests)
+            modules: a list of module names to execute to find tests.
 
-    @classmethod
-    def _discover_tests(cls) -> Iterable[Test]:
-        """
-        Discovers tests in files automatically.
+            filters: regex patterns for test names
+                A match *includes* the test.
 
-        See :envvar:`MODULE` and :envvar:`TESTCASE` for details on how tests are discovered.
         """
-        module_str = os.getenv("MODULE")
-        test_str = os.getenv("TESTCASE")
-
-        if module_str is None:
-            raise ValueError(
-                "Environment variable MODULE, which defines the module(s) to execute, is not defined."
-            )
+        for module_name in modules:
+            self.log.debug("Searching for tests in module %s", module_name)
+            self.log.debug("Python Path: %s", ",".join(sys.path))
+            self.log.debug("PWD: %s", os.getcwd())
+            __import__(module_name)
```

Comment:
This will return the top-level package if `module_name` is multi-level, such as `my_package.my_test_module`.
If we want to replace `_my_import()`, let's use [`importlib.import_module()`](https://docs.python.org/3/library/importlib.html#importlib.import_module).

We also need a testcase for that...

Reply:
We aren't using the return value anymore. Does `__import__` not execute `my_test_module.py` in that case? I'm fine with leaving `_my_import` as is if it is necessary.

Reply:
I think you are right.